File: jLox/src/Stmt.java
import java.util.List;

abstract class Stmt {
    abstract <R> R accept(Visitor<R> visitor);

    interface Visitor<R> {
        R visitPrintStmt(Print stmt);
        R visitExpressionStmt(Expression stmt);
        R visitVarStmt(Var stmt);
        R visitBlockStmt(Block stmt);
        /*
        R visitClassStmt(Class stmt);
        R visitFunctionStmt(Function stmt);
        R visitIfStmt(If stmt);
        R visitReturnStmt(Return stmt);
        R visitWhileStmt(While stmt); */
    }

    static class Expression extends Stmt {
        Expression(Expr expression) {
            this.expression = expression;
        }

        @Override
        <R> R accept(Visitor<R> visitor) {
            return visitor.visitExpressionStmt(this);
        }

        final Expr expression;
    }

    static class Block extends Stmt {
        final List<Stmt> statements;

        Block(List<Stmt> statements) {
            this.statements = statements;
        }

        @Override
        <R> R accept(Visitor<R> visitor) {
            return visitor.visitBlockStmt(this);
        }
    }
    /* Variable Declarations */
    static class Var extends Stmt {
        final Token name;
        final Expr initializer;

        Var(Token name, Expr initializer) {
            this.name = name;
            this.initializer = initializer;
        }

        @Override
        <R> R accept(Visitor<R> visitor) {
            return visitor.visitVarStmt(this);
        }
    }
    static class Print extends Stmt {
        final Expr expression;

        Print(Expr expression) {
            this.expression = expression;
        }

        @Override
        <R> R accept(Visitor<R> visitor) {
            return visitor.visitPrintStmt(this);
        }
    }
}

File: plox/stmt.py
from __future__ import annotations

from dataclasses import dataclass
from tokens import Token, TokenType
from expr import Expr, Variable

class Stmt:
    pass

@dataclass
class Expression(Stmt):
    expression: Expr 

@dataclass
class Return(Stmt):
    keyword: Token 
    value: Expr 

@dataclass
class Function(Stmt):
    name: Token 
    params: list[Token];
    body: list[Stmt]

@dataclass
class While(Stmt):
    condition: Expr 
    body: Stmt 

@dataclass
class If(Stmt):
    condition: Expr 
    thenBranch: Stmt 
    elseBranch: Stmt 

@dataclass
class Block(Stmt):
    statements: list[Stmt]

@dataclass
class Var(Stmt):
    name: Token 
    initializer: Expr 

@dataclass
class Print(Stmt):
    expression: Expr 

@dataclass
class Class(Stmt):
    name: Token 
    methods: list[Function]